jdisc_http_service/src/main/java/com/yahoo/container/logging/LogFileHandler.java

```diff
@@ -110,320 +50,426 @@ private void flushIfOld(long age, TimeUnit unit) {
-            if (currentOut != null) {
-                if (compression == Compression.GZIP) {
-                    long newPos = currentOut.getChannel().position();
-                    if (newPos > lastDropPosition + 102400) {
-                        nativeIO.dropPartialFileFromCache(currentOut.getFD(), lastDropPosition, newPos, true);
-                        lastDropPosition = newPos;
-                    }
-                } else {
-                    currentOut.flush();
-                }
-            }
-        } catch (IOException e) {
-            logger.warning("Failed dropping from cache : " + Exceptions.toMessageString(e));
-        }
+    public void flush() {
+        addOperationAndWait(new Operation<>(Operation.Type.flush));
+    }
+
+    /**
+     * Force file rotation now, independent of schedule.
+     */
+    void rotateNow() {
+        addOperationAndWait(new Operation<>(Operation.Type.rotate));
     }
 
     public void close() {
+        addOperationAndWait(new Operation<>(Operation.Type.close));
+    }
+
+    private void addOperation(Operation<LOGTYPE> op) {
         try {
-            flush();
-            FileOutputStream currentOut = this.currentOutputStream;
-            if (currentOut != null) currentOut.close();
-        } catch (Exception e) {
-            logger.log(Level.WARNING, "Got error while closing log file", e);
+            logThread.logQueue.put(op);
+        } catch (InterruptedException e) {
         }
     }
 
-    private void internalPublish(LOGTYPE r) {
-        // first check to see if new file needed.
-        // if so, use this.internalRotateNow() to do it
-
-        long now = System.currentTimeMillis();
-        if (nextRotationTime <= 0) {
-            nextRotationTime = getNextRotationTime(now); // lazy initialization
-        }
-        if (now > nextRotationTime || currentOutputStream == null) {
-            internalRotateNow();
-        }
+    private void addOperationAndWait(Operation<LOGTYPE> op) {
         try {
-            FileOutputStream out = this.currentOutputStream;
-            logWriter.write(r, out);
-            out.write('\n');
-        } catch (IOException e) {
-            logger.warning("Failed writing log record: " + Exceptions.toMessageString(e));
+            logThread.logQueue.put(op);
+            op.countDownLatch.await();
+        } catch (InterruptedException e) {
         }
     }
 
     /**
-     * Find next rotation after specified time.
-     *
-     * @param now the specified time; if zero, current time is used.
-     * @return the next rotation time
+     * Flushes all queued messages, interrupts the log thread in this and
+     * waits for it to end before returning
      */
-    long getNextRotationTime (long now) {
-        if (now <= 0) {
-            now = System.currentTimeMillis();
+    void shutdown() {
+        logThread.interrupt();
+        try {
+            logThread.join();
+            logThread.executor.shutdown();
```

Comment:
Call `shutdownNow` instead of `shutdown`. Ensure that compression runnable/task handles interruption correctly.